drivers/scmi-msg/scmi_pinctrl.c

```diff
@@ -0,0 +1,810 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2024, EPAM Systems
+ */
+#include <cdefs.h>
+#include <string.h>
+#include <errno.h>
+
+#include <drivers/pinctrl.h>
+#include <drivers/scmi-msg.h>
+#include <drivers/scmi.h>
+#include <lib/utils.h>
+#include <lib/utils_def.h>
+
+#include "common.h"
+#define PP(fmt, ...) \
+	INFO("==== %s %d " fmt "\n", __func__, __LINE__, ##__VA_ARGS__);
+
+#pragma weak pinctrl_get_groups_count
+#pragma weak pinctrl_get_functions_count
+#pragma weak pinctrl_get_pins_count
+#pragma weak pinctrl_get_pins
+#pragma weak pinctrl_get_function_name
+#pragma weak pinctrl_get_group_name
+#pragma weak pinctrl_get_group_pins
+#pragma weak pinctrl_get_function_groups
+#pragma weak pinctrl_request
+#pragma weak pinctrl_free
+
+uint16_t pinctrl_get_groups_count()
+{
+	return 0;
+}
+
+uint16_t pinctrl_get_functions_count()
+{
+	return 0;
+}
+
+uint16_t pinctrl_get_pins_count()
+{
+	return 0;
+}
+
+int pinctrl_get_pins(const struct pinctrl_pin **pins, unsigned *num_pins)
+{
+	return -ENOTSUP;
+}
+
+const char *pinctrl_get_function_name(unsigned selector)
+{
+	return NULL;
+}
+
+const char *pinctrl_get_group_name(unsigned selector)
+{
+	return NULL;
+}
+
+int pinctrl_get_group_pins(unsigned selector, const unsigned **pins,
+			   unsigned *num_pins)
+{
+	return -ENOTSUP;
+}
+
+int pinctrl_get_function_groups(unsigned selector, const int **groups,
+				unsigned *num_groups)
+{
+	return -ENOTSUP;
+}
+
+int pinctrl_request(unsigned offset)
+{
+	return -ENOTSUP;
+}
+
+int pinctrl_free(unsigned offset)
+{
+	return -ENOTSUP;
+}
+
+static bool message_id_is_supported(unsigned int message_id);
+
+static void scmi_pinctrl_report_version(struct scmi_msg *msg)
+{
+	struct scmi_protocol_version_p2a return_values = {
+		.status = SCMI_SUCCESS,
+		.version = SCMI_PROTOCOL_VERSION_PINCTRL,
+	};
+
+	PP("scmi: Process pinctrl protocol_version\n");
+
+	if (msg->in_size != 0U) {
+		PP("");
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	scmi_write_response(msg, &return_values, sizeof(return_values));
+}
+
+static void scmi_pinctrl_report_attributes(struct scmi_msg *msg)
+{
+	uint16_t nr_functions = pinctrl_get_functions_count();
+	uint16_t nr_groups = pinctrl_get_groups_count();
+	unsigned nr_pins = pinctrl_get_pins_count();
+	struct scmi_pinctrl_proro_attrs_a2p {
+		int32_t status;
+		uint32_t attrs_low;
+		uint32_t attrs_high;
+	} return_values = {.status = SCMI_SUCCESS};
+
+	if (msg->in_size != 0U) {
+		scmi_status_response(msg, SCMI_PROTOCOL_ERROR);
+		return;
+	}
+
+	return_values.attrs_low =
+		FLD(GENMASK(31, 16), nr_groups) | FLD(GENMASK(15, 0), nr_pins);
+	return_values.attrs_high = FLD(GENMASK(15, 0), nr_functions);
+
+	INFO("scmi: pinctrl protocol_attrs group_max = %d, func_max = "
```

Comment:
this probably needs to be verbose as well

Reply:
fixed